app/common/console_output.py

```diff
@@ -1,64 +1,81 @@
         """
-        self.path = path
+        self._file = file
 
-    def segment(self, max_lines=50, offset_line=None):
+    def segment(self, max_lines: int=50, offset_line: Optional[int]=None):
         """
-        Return a segment of the console output.
+        Return a segment of the console output. Note this closes the file handle so can only be called once
+        per ConsoleOutput object.
 
-        :type max_lines: int
+        :param max_lines: The maximum number of lines of output to return
         :param offset_line: The starting line number in the console output from which the segment should
             return content for. If set to None, then reads content starting from the end.
-        :type offset_line: int | None
-        :rtype: ConsoleOutputSegment
         """
         if offset_line is None:
             return self._parse_from_end(max_lines)
         else:
             return self._parse_from_offset(max_lines, offset_line)
 
-    def _parse_from_offset(self, max_lines, offset_line):
+    def _parse_from_offset(self, max_lines: int, offset_line: int) -> ConsoleOutputSegment:
         """
         Return up to max_lines of output starting from line number offset_line. If max_lines + offset_line
         doesn't exist on this console output, then return everything beyond offset_line.
-
-        :type max_lines: int
-        :type offset_line: int
-        :rtype: ConsoleOutputSegment
         """
         total_lines = 0
         output_lines = 0
         console_output = []
 
-        with open(self.path, 'r', encoding='utf-8', errors='replace') as f:
+        with self._file as f:
             # Iterate up to the index offset_line
             for i in range(0, offset_line):
-                # This is an error, meaning that there aren't even offset_line+1 lines in self.path.
-                if f.readline() == '':
-                    raise ValueError('offset: {} is higher than the total number of lines in file {}'.format(
-                        offset_line, self.path))
-
+                # This is an error, meaning that there aren't even offset_line+1 lines in the file.
+                if len(f.readline()) == 0:
+                    raise BadConsoleOutputRequestError(
+                        'offset {} is higher than the total number of lines: {}'.format(offset_line, total_lines))
                 total_lines += 1
 
             # Retrieve the console_output just between offset_line and offset_line + max_lines
             for i in range(offset_line, offset_line + max_lines):
                 line = f.readline()
 
                 # We have reached the end of the file, or a line that has not finished being written to.
-                if line == '' or not line.endswith("\n"):
+                if not line.endswith(b'\n'):
                     break
 
-                console_output.append(line)
+                console_output.append(line.decode(encoding='utf-8', errors='replace'))
```

Comment:
This is actually equivalent to what we were doing before. Previously this class would open the file in text mode so these params were passed with the `open()` call. Now it's opened in binary mode (to be compatible with how zip files are read) so the decoding happens here.

I don't remember if we added errors=replace in response to some bug, but I'd guess that was the case. We could probably dig it up in the git history.